rust/operator-binary/src/config/jvm.rs

```diff
@@ -0,0 +1,205 @@
+    role_utils::JvmArgumentOverrides,
+};
+
+const JVM_HEAP_FACTOR: f32 = 0.8;
+
+#[derive(Snafu, Debug)]
+pub enum Error {
+    #[snafu(display("invalid java heap config for {role:?}"))]
+    InvalidJavaHeapConfig {
+        source: stackable_operator::memory::Error,
+        role: String,
+    },
+
+    #[snafu(display("failed to merge jvm argument overrides"))]
+    MergeJvmArgumentOverrides { source: stackable_hdfs_crd::Error },
+}
+
+// All init or sidecar containers must have access to the following settings.
+// As the Prometheus metric emitter is not part of this config it's safe to use for hdfs cli tools as well.
+// This will not only enable the init containers to work, but also the user to run e.g.
+// `bin/hdfs dfs -ls /` without getting `Caused by: java.lang.IllegalArgumentException: KrbException: Cannot locate default realm`
+// because the `-Djava.security.krb5.conf` setting is missing
+pub fn construct_global_jvm_args(kerberos_enabled: bool) -> String {
+    let mut jvm_args = Vec::new();
+
+    if kerberos_enabled {
+        jvm_args.push("-Djava.security.krb5.conf=/stackable/kerberos/krb5.conf".to_owned());
+    }
+
+    // We do *not* add user overrides to the global JVM args, but only the role specific JVm arguments.
+    // This allows users to configure stuff for the server (probably what they want to do), without
+    // also influencing e.g. startup scripts.
+    //
+    // However, this is just an assumptions. If it is wrong users can still envOverride the global
+    // JVM args.
+    //
+    // Please feel absolutely free to change this behavior!
+    jvm_args.join(" ")
+}
+
+pub fn construct_role_specific_jvm_args(
+    hdfs: &HdfsCluster,
+    hdfs_role: &HdfsRole,
+    role_group: &str,
+    kerberos_enabled: bool,
+    resources: Option<&ResourceRequirements>,
+    config_dir: &str,
+    metrics_port: u16,
+) -> Result<String, Error> {
+    let mut jvm_args = Vec::new();
+
+    if let Some(memory_limit) = resources.and_then(|r| r.limits.as_ref()?.get("memory")) {
+        let memory_limit = MemoryQuantity::try_from(memory_limit).with_context(|_| {
+            InvalidJavaHeapConfigSnafu {
+                role: hdfs_role.to_string(),
+            }
+        })?;
+        let heap = memory_limit.scale_to(BinaryMultiple::Mebi) * JVM_HEAP_FACTOR;
+        let heap = heap
+            .format_for_java()
+            .with_context(|_| InvalidJavaHeapConfigSnafu {
+                role: hdfs_role.to_string(),
+            })?;
+
+        jvm_args.push(format!("-Xms{heap}"));
+        jvm_args.push(format!("-Xmx{heap}"));
+    }
+
+    jvm_args.extend([
+        format!("-Djava.security.properties={config_dir}/{JVM_SECURITY_PROPERTIES_FILE}"),
+        format!("-javaagent:/stackable/jmx/jmx_prometheus_javaagent.jar={metrics_port}:/stackable/jmx/{hdfs_role}.yaml")
+    ]);
+    if kerberos_enabled {
+        jvm_args.push("-Djava.security.krb5.conf=/stackable/kerberos/krb5.conf".to_string());
+    }
```

Comment:
Might want to use `construct_global_jvm_args` here or at the start? Otherwise you can move `construct_global_jvm_args`down to the tests section.

Reply:
I *think* it's mostly a coincidence that the global args match this specific arg.
There might be a case in the future that the cli tools/zkfc/format- namenodes need some different arguments than the server.
Or that Kerberos in the server configures more stuff than the cli tools.

Reply:
`construct_global_jvm_args` ends up in `HADOOP_OPTS`, so not a test-only thing.
`construct_role_specific_jvm_args` ends up in e.g. `HDFS_NAMENODE_OPTS`